Sources/Conduit/Networking/Data/DataDownloader.swift

```diff
@@ -0,0 +1,141 @@
+            self.data = data
+            self.error = error
+            self.urlResponse = urlResponse
+            self.isFromCache = isFromCache
+        }
+    }
+
+    /// A closure that fires upon data fetch success/failure
+    public typealias CompletionHandler = (Response) -> Void
+
+    private var cache: URLDataCache
+    private let sessionClient: URLSessionClientType
+    private var sessionProxyMap: [String: SessionTaskProxyType] = [:]
+    private var completionHandlerMap: [String: [CompletionHandler]] = [:]
+    private let completionQueue: OperationQueue?
+    private let serialQueue = DispatchQueue(
+        label: "com.mindbodyonline.Conduit.DataDownloader-\(UUID().uuidString)"
+    )
+
+    /// Initializes a new DataDownloader
+    /// - Parameters:
+    ///   - cache: The data cache in which to store downloaded data
+    ///   - sessionClient: The URLSessionClient to be used to download data
+    ///   - completionQueue: An optional operation queue for completion callback
+    public init(cache: URLDataCache,
+                sessionClient: URLSessionClientType = URLSessionClient(),
+                completionQueue: OperationQueue? = nil) {
+        self.cache = cache
+        self.sessionClient = sessionClient
+        self.completionQueue = completionQueue
+    }
+
+    /// Downloads data or retrieves it from the cache if previously downloaded.
+    /// - Parameters:
+    ///     - request: The request for the data
+    /// - Returns: A concrete SessionTaskProxyType
+    @discardableResult
+    public func downloadData(for request: URLRequest, completion: @escaping CompletionHandler) -> SessionTaskProxyType? {
+        var proxy: SessionTaskProxyType?
+        let completionQueue = self.completionQueue ?? .current ?? .main
+
+        serialQueue.sync { [weak self] in
+            guard let `self` = self else {
+                return
+            }
+
+            if let data = self.cache.data(for: request) {
+                let response = Response(data: data, error: nil, urlResponse: nil, isFromCache: true)
+                completion(response)
+                return
+            }
+
+            guard let cacheIdentifier = self.cache.cacheIdentifier(for: request) else {
+                let response = Response(data: nil,
+                                        error: DataDownloaderError.invalidRequest,
+                                        urlResponse: nil,
+                                        isFromCache: false)
+                completion(response)
+                return
+            }
+
+            self.register(completionHandler: completion, for: cacheIdentifier)
+
+            if let sessionTaskProxy = self.sessionProxyMap[cacheIdentifier] {
+                proxy = sessionTaskProxy
+                return
+            }
+
+            // Strongly capture self within the completion handler to ensure
+            // DataDownloader is persisted long enough to respond
+            let strongSelf = self
```

Comment:
with the above guard statement (line 79) defined as it is `self` here is non-optional unless I'm misreading the scopes

Reply:
Still don't think this is needed.